src/arrow2/src/array/mod.rs

```diff
@@ -633,15 +636,27 @@ macro_rules! impl_common_array {
 
         fn change_type(&mut self, data_type: DataType) {
             if data_type.to_physical_type() != self.data_type().to_physical_type() {
-                panic!(
-                    "Converting array with logical type {:?} to logical type {:?} failed, physical types do not match: {:?} -> {:?}",
-                    self.data_type(),
-                    data_type,
-                    self.data_type().to_physical_type(),
-                    data_type.to_physical_type(),
-                );
+                let from =                     format!("{:#?}", self.data_type());
+                let to = format!("{:#?}", data_type);
+
+
+                let diff = similar::TextDiff::from_lines(&from, &to);
+
+                let diff = diff
+                    .unified_diff();
+
+                panic!("{diff}");
```

Comment:
Could we write a more descriptive error message to go with the diff of the types?

Reply:
believe I changed this